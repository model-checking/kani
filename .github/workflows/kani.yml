# Copyright Kani Contributors
# SPDX-License-Identifier: Apache-2.0 OR MIT
name: Kani CI
on:
  pull_request:
  push:
    # Not just any push, as that includes tags.
    # We don't want to re-trigger this workflow when tagging an existing commit.
    branches:
      - '**'

env:
  RUST_BACKTRACE: 1

jobs:
  regression:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [macos-12, ubuntu-20.04, ubuntu-22.04]
    steps:
      - name: Checkout Kani
        uses: actions/checkout@v4

      - name: Setup Kani Dependencies
        uses: ./.github/actions/setup
        with:
          os: ${{ matrix.os }}

      - name: Build Kani
        run: cargo build-dev

      - name: Execute Kani regression
        run: ./scripts/kani-regression.sh

  write-json-symtab-regression:
    runs-on: ubuntu-20.04
    steps:
      - name: Checkout Kani
        uses: actions/checkout@v4

      - name: Setup Kani Dependencies
        uses: ./.github/actions/setup
        with:
          os: ubuntu-20.04

      - name: Build Kani
        run: cargo build-dev -- --features write_json_symtab

      - name: Run tests
        run: |
          cargo run -p compiletest --quiet -- --suite kani --mode kani --quiet --no-fail-fast
          cargo run -p compiletest --quiet -- --suite expected --mode expected --quiet --no-fail-fast
          cargo run -p compiletest --quiet -- --suite cargo-kani --mode cargo-kani --quiet --no-fail-fast


  benchcomp-tests:
    runs-on: ubuntu-20.04
    steps:
      - name: Checkout Kani
        uses: actions/checkout@v4

      - name: Install benchcomp dependencies
        run: |
          sudo apt-get update
          sudo apt-get install -y python3-pip
          pushd tools/benchcomp && pip3 install -r requirements.txt

      - name: Setup Kani Dependencies
        uses: ./.github/actions/setup
        with:
          os: ubuntu-20.04

      - name: Build Kani using release mode
        run: cargo build-dev -- --release

      - name: Run benchcomp unit and regression tests
        run: pushd tools/benchcomp && PATH=$(realpath ../../scripts):$PATH test/run

  perf:
    runs-on: ubuntu-20.04
    steps:
      - name: Checkout Kani
        uses: actions/checkout@v4

      - name: Setup Kani Dependencies
        uses: ./.github/actions/setup
        with:
          os: ubuntu-20.04

      - name: Build Kani using release mode
        run: cargo build-dev -- --release

      - name: Execute Kani performance tests
        run: ./scripts/kani-perf.sh
        env:
          RUST_TEST_THREADS: 1

  bookrunner:
    runs-on: ubuntu-20.04
    permissions:
      contents: write
    steps:
      - name: Checkout Kani
        uses: actions/checkout@v4

      - name: Setup Kani Dependencies
        uses: ./.github/actions/setup
        with:
          os: ubuntu-20.04

      - name: Build Kani
        run: cargo build-dev

      - name: Install book runner dependencies
        run: ./scripts/setup/install_bookrunner_deps.sh

      - name: Generate book runner report
        run: cargo run -p bookrunner
        env:
          DOC_RUST_LANG_ORG_CHANNEL: nightly

      - name: Print book runner text results
        run: cat build/output/latest/html/bookrunner.txt

      - name: Print book runner failures grouped by stage
        run: python3 scripts/ci/bookrunner_failures_by_stage.py build/output/latest/html/index.html

      - name: Detect unexpected book runner failures
        run: ./scripts/ci/detect_bookrunner_failures.sh build/output/latest/html/bookrunner.txt

      - name: Install book dependencies
        run: ./scripts/setup/ubuntu/install_doc_deps.sh

      # On one OS only, build the documentation, too.
      - name: Build Documentation
        run: ./scripts/build-docs.sh

      # When we're pushed to main branch, only then actually publish the docs.
      - name: Publish Documentation
        if: ${{ github.event_name == 'push' && startsWith('refs/heads/main', github.ref) }}
        uses: JamesIves/github-pages-deploy-action@v4
        with:
          branch: gh-pages
          folder: docs/book/
          single-commit: true
