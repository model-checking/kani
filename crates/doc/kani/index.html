<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `kani` crate."><title>kani - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (1d23d0680 2025-10-22)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Crate kani</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../kani/index.html">kani</a><span class="version">0.65.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#reexports">Crate Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li><li><a href="#derives" title="Derive Macros">Derive Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>kani</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/kani/lib.rs.html#5-93">Source</a> </span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><dl class="item-table reexports"><dt id="reexport.Invariant"><code>pub use invariant::<a class="trait" href="invariant/trait.Invariant.html" title="trait kani::invariant::Invariant">Invariant</a>;</code></dt><dt id="reexport.RoundRobin"><code>pub use futures::<a class="struct" href="futures/struct.RoundRobin.html" title="struct kani::futures::RoundRobin">RoundRobin</a>;</code></dt><dt id="reexport.block_on"><code>pub use futures::<a class="fn" href="futures/fn.block_on.html" title="fn kani::futures::block_on">block_on</a>;</code></dt><dt id="reexport.block_on_with_spawn"><code>pub use futures::<a class="fn" href="futures/fn.block_on_with_spawn.html" title="fn kani::futures::block_on_with_spawn">block_on_with_spawn</a>;</code></dt><dt id="reexport.spawn"><code>pub use futures::<a class="fn" href="futures/fn.spawn.html" title="fn kani::futures::spawn">spawn</a>;</code></dt><dt id="reexport.yield_now"><code>pub use futures::<a class="fn" href="futures/fn.yield_now.html" title="fn kani::futures::yield_now">yield_now</a>;</code></dt></dl><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="arbitrary/index.html" title="mod kani::arbitrary">arbitrary</a></dt><dd>This module introduces the <code>Arbitrary</code> trait as well as implementation for
primitive types and other std containers.</dd><dt><a class="mod" href="bounded_arbitrary/index.html" title="mod kani::bounded_arbitrary">bounded_<wbr>arbitrary</a></dt><dd>This module introduces implementations for some std containers.</dd><dt><a class="mod" href="contracts/index.html" title="mod kani::contracts">contracts</a></dt><dd>Kani implementation of function contracts.</dd><dt><a class="mod" href="float/index.html" title="mod kani::float">float</a></dt><dd>This module contains functions useful for float-related checks</dd><dt><a class="mod" href="futures/index.html" title="mod kani::futures">futures</a></dt><dd>This module contains functions to work with futures (and async/.await) in Kani.</dd><dt><a class="mod" href="invariant/index.html" title="mod kani::invariant">invariant</a></dt><dd>This module introduces the <code>Invariant</code> trait as well as its implementation
for primitive types.</dd><dt><a class="mod" href="iter/index.html" title="mod kani::iter">iter</a></dt><dd>This macro generates implementations of the <code>KaniIntoIter</code> trait for various common types that are used in <code>for</code> loops.
We use this trait to overwrite the Rust IntoIter trait to reduce call stacks and avoid complicated loop invariant specifications,
while maintaining the semantics of the loop.</dd><dt><a class="mod" href="mem/index.html" title="mod kani::mem">mem</a></dt><dd>This module contains functions useful for checking unsafe memory access.</dd><dt><a class="mod" href="shadow/index.html" title="mod kani::shadow">shadow</a></dt><dd>This module contains an API for shadow memory.
Shadow memory is a mechanism by which we can store metadata on memory
locations, e.g. whether a memory location is initialized.</dd><dt><a class="mod" href="slice/index.html" title="mod kani::slice">slice</a></dt><dt><a class="mod" href="vec/index.html" title="mod kani::vec">vec</a></dt></dl><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><dl class="item-table"><dt><a class="macro" href="macro.arbitrary_tuple.html" title="macro kani::arbitrary_tuple">arbitrary_<wbr>tuple</a></dt><dd>This macro implements <code>kani::Arbitrary</code> on a tuple whose elements
already implement <code>kani::Arbitrary</code> by running <code>kani::any()</code> on
each index of the tuple.</dd><dt><a class="macro" href="macro.cover.html" title="macro kani::cover">cover</a></dt><dt><a class="macro" href="macro.exists.html" title="macro kani::exists">exists</a></dt><dt><a class="macro" href="macro.forall.html" title="macro kani::forall">forall</a></dt><dt><a class="macro" href="macro.generate_arbitrary.html" title="macro kani::generate_arbitrary">generate_<wbr>arbitrary</a></dt><dt><a class="macro" href="macro.generate_bounded_arbitrary.html" title="macro kani::generate_bounded_arbitrary">generate_<wbr>bounded_<wbr>arbitrary</a></dt><dt><a class="macro" href="macro.generate_float.html" title="macro kani::generate_float">generate_<wbr>float</a></dt><dt><a class="macro" href="macro.generate_iter.html" title="macro kani::generate_iter">generate_<wbr>iter</a></dt><dt><a class="macro" href="macro.generate_models.html" title="macro kani::generate_models">generate_<wbr>models</a></dt><dt><a class="macro" href="macro.implies.html" title="macro kani::implies">implies</a></dt><dd><code>implies!(premise =&gt; conclusion)</code> means that if the <code>premise</code> is true, so
must be the <code>conclusion</code>.</dd><dt><a class="macro" href="macro.kani_intrinsics.html" title="macro kani::kani_intrinsics">kani_<wbr>intrinsics</a></dt><dd>Kani intrinsics contains the public APIs used by users to verify their harnesses.
This macro is a part of kani_core as that allows us to verify even libraries that are no_core
such as core in rust’s std library itself.</dd><dt><a class="macro" href="macro.kani_lib.html" title="macro kani::kani_lib">kani_<wbr>lib</a></dt><dd>Users should only need to invoke this.</dd><dt><a class="macro" href="macro.kani_mem.html" title="macro kani::kani_mem">kani_<wbr>mem</a></dt><dt><a class="macro" href="macro.kani_mem_init.html" title="macro kani::kani_mem_init">kani_<wbr>mem_<wbr>init</a></dt><dt><a class="macro" href="macro.ptr_generator.html" title="macro kani::ptr_generator">ptr_<wbr>generator</a></dt><dt><a class="macro" href="macro.ptr_generator_fn.html" title="macro kani::ptr_generator_fn">ptr_<wbr>generator_<wbr>fn</a></dt><dt><a class="macro" href="macro.slice_generator.html" title="macro kani::slice_generator">slice_<wbr>generator</a></dt></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArbitraryPointer.html" title="struct kani::ArbitraryPointer">Arbitrary<wbr>Pointer</a></dt><dd>Holds information about a pointer that is generated non-deterministically.</dd><dt><a class="struct" href="struct.KaniChainIter.html" title="struct kani::KaniChainIter">Kani<wbr>Chain<wbr>Iter</a></dt><dt><a class="struct" href="struct.KaniEnumerateIter.html" title="struct kani::KaniEnumerateIter">Kani<wbr>Enumerate<wbr>Iter</a></dt><dt><a class="struct" href="struct.KaniMapIter.html" title="struct kani::KaniMapIter">Kani<wbr>MapIter</a></dt><dt><a class="struct" href="struct.KaniPtrIter.html" title="struct kani::KaniPtrIter">Kani<wbr>PtrIter</a></dt><dt><a class="struct" href="struct.KaniRefIter.html" title="struct kani::KaniRefIter">Kani<wbr>RefIter</a></dt><dt><a class="struct" href="struct.KaniRevIter.html" title="struct kani::KaniRevIter">Kani<wbr>RevIter</a></dt><dt><a class="struct" href="struct.KaniStepBy.html" title="struct kani::KaniStepBy">Kani<wbr>Step<wbr>By</a></dt><dt><a class="struct" href="struct.KaniTakeIter.html" title="struct kani::KaniTakeIter">Kani<wbr>Take<wbr>Iter</a></dt><dt><a class="struct" href="struct.KaniZipIter.html" title="struct kani::KaniZipIter">Kani<wbr>ZipIter</a></dt><dt><a class="struct" href="struct.PointerGenerator.html" title="struct kani::PointerGenerator">Pointer<wbr>Generator</a></dt><dd>Pointer generator that can be used to generate arbitrary pointers.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.AllocationStatus.html" title="enum kani::AllocationStatus">Allocation<wbr>Status</a></dt><dd>Enumeration with the cases currently covered by the pointer generator.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.Arbitrary.html" title="trait kani::Arbitrary">Arbitrary</a></dt><dt><a class="trait" href="trait.BoundedArbitrary.html" title="trait kani::BoundedArbitrary">Bounded<wbr>Arbitrary</a></dt><dt><a class="trait" href="trait.KaniIntoIter.html" title="trait kani::KaniIntoIter">Kani<wbr>Into<wbr>Iter</a></dt><dt><a class="trait" href="trait.KaniIter.html" title="trait kani::KaniIter">Kani<wbr>Iter</a></dt></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.any.html" title="fn kani::any">any</a></dt><dd>This creates an symbolic <em>valid</em> value of type <code>T</code>. You can assign the return value of this
function to a variable that you want to make symbolic.</dd><dt><a class="fn" href="fn.any_where.html" title="fn kani::any_where">any_<wbr>where</a></dt><dd>This creates a symbolic <em>valid</em> value of type <code>T</code>.
The value is constrained to be a value accepted by the predicate passed to the filter.
You can assign the return value of this function to a variable that you want to make symbolic.</dd><dt><a class="fn" href="fn.assert.html" title="fn kani::assert">assert</a></dt><dd>Creates an assertion of the specified condition and message.</dd><dt><a class="fn" href="fn.assume.html" title="fn kani::assume">assume</a></dt><dd>Creates an assumption that will be valid after this statement run. Note that the assumption
will only be applied for paths that follow the assumption. If the assumption doesn’t hold, the
program will exit successfully.</dd><dt><a class="fn" href="fn.bounded_any.html" title="fn kani::bounded_any">bounded_<wbr>any</a></dt><dd>Creates a symbolic value <em>bounded</em> by <code>N</code>. Bounded means <code>|T| &lt;= N</code>. The type
implementing BoundedArbitrary decides exactly what size means for them.</dd><dt><a class="fn" href="fn.concrete_playback_run.html" title="fn kani::concrete_playback_run">concrete_<wbr>playback_<wbr>run</a></dt><dd>NOP <code>concrete_playback</code> for type checking during verification mode.</dd><dt><a class="fn" href="fn.cover.html" title="fn kani::cover">cover</a></dt><dd>Creates a cover property with the specified condition and message.</dd><dt><a class="fn" href="fn.pointer_generator.html" title="fn kani::pointer_generator">pointer_<wbr>generator</a></dt><dd>Create a pointer generator that fits at least <code>N</code> elements of type <code>T</code>.</dd></dl><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><dl class="item-table"><dt><a class="attr" href="attr.ensures.html" title="attr kani::ensures">ensures</a></dt><dd>Add a postcondition to this function.</dd><dt><a class="attr" href="attr.loop_invariant.html" title="attr kani::loop_invariant">loop_<wbr>invariant</a></dt><dd>Add a loop invariant to this loop.</dd><dt><a class="attr" href="attr.loop_modifies.html" title="attr kani::loop_modifies">loop_<wbr>modifies</a></dt><dt><a class="attr" href="attr.modifies.html" title="attr kani::modifies">modifies</a></dt><dd>Declaration of an explicit write-set for the annotated function.</dd><dt><a class="attr" href="attr.proof.html" title="attr kani::proof">proof</a></dt><dd>Marks a Kani proof harness</dd><dt><a class="attr" href="attr.proof_for_contract.html" title="attr kani::proof_for_contract">proof_<wbr>for_<wbr>contract</a></dt><dd>Designates this function as a harness to check a function contract.</dd><dt><a class="attr" href="attr.recursion.html" title="attr kani::recursion">recursion</a></dt><dd>Specifies that a function contains recursion for contract instrumentation.**</dd><dt><a class="attr" href="attr.requires.html" title="attr kani::requires">requires</a></dt><dd>Add a precondition to this function.</dd><dt><a class="attr" href="attr.should_panic.html" title="attr kani::should_panic">should_<wbr>panic</a></dt><dd>Specifies that a proof harness is expected to panic.**</dd><dt><a class="attr" href="attr.solver.html" title="attr kani::solver">solver</a></dt><dd>Select the SAT solver to use with CBMC for this harness</dd><dt><a class="attr" href="attr.stub.html" title="attr kani::stub">stub</a></dt><dd>Specify a function/method stub pair to use for proof harness</dd><dt><a class="attr" href="attr.stub_verified.html" title="attr kani::stub_verified">stub_<wbr>verified</a></dt><dd><code>stub_verified(TARGET)</code> is a harness attribute (to be used on
<a href="attr.proof.html" title="attr kani::proof"><code>proof</code></a> or <a href="attr.proof_for_contract.html" title="attr kani::proof_for_contract"><code>proof_for_contract</code></a>
function) that replaces all occurrences of <code>TARGET</code> reachable from this
harness with a stub generated from the contract on <code>TARGET</code>.</dd><dt><a class="attr" href="attr.unwind.html" title="attr kani::unwind">unwind</a></dt><dd>Set Loop unwind limit for proof harnesses
The attribute <code>#[kani::unwind(arg)]</code> can only be called alongside <code>#[kani::proof]</code>.
arg - Takes in a integer value (u32) that represents the unwind value for the harness.</dd></dl><h2 id="derives" class="section-header">Derive Macros<a href="#derives" class="anchor">§</a></h2><dl class="item-table"><dt><a class="derive" href="derive.Arbitrary.html" title="derive kani::Arbitrary">Arbitrary</a></dt><dd>Allow users to auto generate <code>Arbitrary</code> implementations by using
<code>#[derive(Arbitrary)]</code> macro.</dd><dt><a class="derive" href="derive.BoundedArbitrary.html" title="derive kani::BoundedArbitrary">Bounded<wbr>Arbitrary</a></dt><dd>Allow users to generate <code>BoundedArbitrary</code> implementations by using the
<code>#[derive(BoundedArbitrary)]</code> macro.</dd><dt><a class="derive" href="derive.Invariant.html" title="derive kani::Invariant">Invariant</a></dt><dd>Allow users to auto generate <code>Invariant</code> implementations by using
<code>#[derive(Invariant)]</code> macro.</dd></dl></section></div></main></body></html>