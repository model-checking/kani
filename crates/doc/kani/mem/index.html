<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This module contains functions useful for checking unsafe memory access."><title>kani::mem - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (1d23d0680 2025-10-22)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module mem</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../kani/index.html">kani</a><span class="version">0.65.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module mem</a></h2><h3><a href="#functions">Module Items</a></h3><ul class="block"><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate kani</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">kani</a></div><h1>Module <span>mem</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/kani/lib.rs.html#57">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This module contains functions useful for checking unsafe memory access.</p>
<p>Given the following validity rules provided in the Rust documentation:
<a href="https://doc.rust-lang.org/std/ptr/index.html">https://doc.rust-lang.org/std/ptr/index.html</a> (accessed May 20th, 2025)</p>
<ol>
<li>For memory accesses of size zero, every pointer is valid, including the null pointer.
The following points are only concerned with non-zero-sized accesses.</li>
<li>A null pointer is never valid.</li>
<li>For a pointer to be valid, it is necessary, but not always sufficient, that the pointer
be dereferenceable: the memory range of the given size starting at the pointer must all be
within the bounds of a single allocated object. Note that in Rust, every (stack-allocated)
variable is considered a separate allocated object.</li>
<li>All accesses performed by functions in this module are non-atomic in the sense of atomic
operations used to synchronize between threads.
This means it is undefined behavior to perform two concurrent accesses to the same location
from different threads unless both accesses only read from memory.
Notice that this explicitly includes <code>read_volatile</code> and <code>write_volatile</code>:
Volatile accesses cannot be used for inter-thread synchronization.</li>
<li>The result of casting a reference to a pointer is valid for as long as the underlying
object is live and no reference (just raw pointers) is used to access the same memory.
That is, reference and pointer accesses cannot be interleaved.</li>
</ol>
<p>Kani is able to verify #1, #2, and #3 today.</p>
</div></details><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="fn" href="fn.can_dereference.html" title="fn kani::mem::can_dereference">can_<wbr>dereference</a></dt><dd>Checks that pointer <code>ptr</code> point to a valid value of type <code>T</code>.</dd><dt><a class="fn" href="fn.can_read_unaligned.html" title="fn kani::mem::can_read_unaligned">can_<wbr>read_<wbr>unaligned</a></dt><dd>Checks that pointer <code>ptr</code> point to a valid value of type <code>T</code>.</dd><dt><a class="fn" href="fn.can_write.html" title="fn kani::mem::can_write">can_<wbr>write</a></dt><dd>Check if the pointer is valid for write access according to <a href="index.html" title="mod kani::mem">crate::mem</a> conditions 1, 2
and 3.</dd><dt><a class="fn" href="fn.can_write_unaligned.html" title="fn kani::mem::can_write_unaligned">can_<wbr>write_<wbr>unaligned</a></dt><dd>Check if the pointer is valid for unaligned write access according to <a href="index.html" title="mod kani::mem">crate::mem</a> conditions
1, 2 and 3.</dd><dt><a class="fn" href="fn.checked_align_of_raw.html" title="fn kani::mem::checked_align_of_raw">checked_<wbr>align_<wbr>of_<wbr>raw</a></dt><dd>Compute the size of the val pointed to if safe.</dd><dt><a class="fn" href="fn.checked_size_of_raw.html" title="fn kani::mem::checked_size_of_raw">checked_<wbr>size_<wbr>of_<wbr>raw</a></dt><dd>Compute the size of the val pointed to if it is safe to do so.</dd><dt><a class="fn" href="fn.is_inbounds.html" title="fn kani::mem::is_inbounds">is_<wbr>inbounds</a></dt><dd>Checks that <code>ptr</code> points to an allocation that can hold data of size calculated from <code>T</code>.</dd><dt><a class="fn" href="fn.same_allocation.html" title="fn kani::mem::same_allocation">same_<wbr>allocation</a></dt><dd>Check if two pointers points to the same allocated object, and that both pointers
are in bounds of that object.</dd></dl></section></div></main></body></html>