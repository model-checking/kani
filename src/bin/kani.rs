// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT

use anyhow::Result;

fn main() -> Result<()> {
    kani_verifier::proxy("kani")
}
