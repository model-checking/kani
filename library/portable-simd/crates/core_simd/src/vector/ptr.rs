//! Private implementation details of public gather/scatter APIs.
use crate::simd::{LaneCount, Simd, SupportedLaneCount};
use core::mem;

/// A vector of *const T.
#[derive(Debug, Copy, Clone)]
#[repr(simd)]
pub(crate) struct SimdConstPtr<T, const LANES: usize>([*const T; LANES]);

impl<T, const LANES: usize> SimdConstPtr<T, LANES>
where
    LaneCount<LANES>: SupportedLaneCount,
    T: Sized,
{
    #[inline]
    #[must_use]
    pub fn splat(ptr: *const T) -> Self {
        Self([ptr; LANES])
    }

    #[inline]
    #[must_use]
    pub fn wrapping_add(self, addend: Simd<usize, LANES>) -> Self {
        unsafe {
            let x: Simd<usize, LANES> = mem::transmute_copy(&self);
            mem::transmute_copy(&{ x + (addend * Simd::splat(mem::size_of::<T>())) })
        }
    }
}

/// A vector of *mut T. Be very careful around potential aliasing.
#[derive(Debug, Copy, Clone)]
#[repr(simd)]
pub(crate) struct SimdMutPtr<T, const LANES: usize>([*mut T; LANES]);

impl<T, const LANES: usize> SimdMutPtr<T, LANES>
where
    LaneCount<LANES>: SupportedLaneCount,
    T: Sized,
{
    #[inline]
    #[must_use]
    pub fn splat(ptr: *mut T) -> Self {
        Self([ptr; LANES])
    }

    #[inline]
    #[must_use]
    pub fn wrapping_add(self, addend: Simd<usize, LANES>) -> Self {
        unsafe {
            let x: Simd<usize, LANES> = mem::transmute_copy(&self);
            mem::transmute_copy(&{ x + (addend * Simd::splat(mem::size_of::<T>())) })
        }
    }
}
