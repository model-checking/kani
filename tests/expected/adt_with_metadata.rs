// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Z mem-predicates
//! Check that Kani's memory predicates work for ADT with metadata.
//! TODO: Add test with packed layout and with ZST converted to dyn T.
#![feature(ptr_metadata)]

extern crate kani;

use kani::mem::{can_dereference, can_write};

#[derive(Clone, Copy, kani::Arbitrary)]
struct Wrapper<T: ?Sized> {
    _size: usize,
    _value: T,
}

mod valid_access {
    use super::*;
    use std::ptr;
    #[kani::proof]
    pub fn check_valid_dyn_ptr() {
        let mut var: Wrapper<u64> = kani::any();
        let fat_ptr: *mut Wrapper<dyn PartialEq<u64>> = &mut var as *mut _;
        assert!(can_write(fat_ptr));
    }

    /// In this harness, we cast `Wrapper<[u64;4]>` to `*const [u64]` with length 5.
    /// Since `Wrapper` has an extra usize, it is safe to dereference the new pointer.
    #[kani::proof]
    pub fn check_arbitrary_metadata_is_sound() {
        let mut var: Wrapper<[u64; 4]> = kani::any();
        let fat_ptr: *mut Wrapper<[u64]> = &mut var as *mut _;
        let (thin_ptr, size) = fat_ptr.to_raw_parts();
        let new_size: usize = size + 1;
        let new_ptr: *const [u64] = ptr::from_raw_parts(thin_ptr, new_size);
        assert!(can_dereference(new_ptr));
    }
}

mod invalid_access {
    use super::*;
    use std::ptr;
    #[kani::proof]
    pub fn check_invalid_dyn_ptr() {
        unsafe fn new_dead_ptr<T>(val: T) -> *const T {
            let local = val;
            &local as *const _
        }

        let raw_ptr: *const Wrapper<dyn PartialEq<u8>> =
            unsafe { new_dead_ptr::<Wrapper<u8>>(kani::any()) };
        assert!(can_dereference(raw_ptr));
    }

    /// Check that `can_dereference` correctly identifies cases where creating a pointer from raw
    /// parts with an artificial size is unsafe.
    #[kani::proof]
    pub fn check_arbitrary_metadata() {
        let mut var: Wrapper<[u64; 4]> = kani::any();
        let fat_ptr: *mut Wrapper<[u64]> = &mut var as *mut _;
        let (thin_ptr, size) = fat_ptr.to_raw_parts();
        let new_size: usize = kani::any();
        let new_ptr: *const Wrapper<[u64]> = ptr::from_raw_parts(thin_ptr, new_size);
        if new_size <= size {
            assert!(can_dereference(new_ptr));
        } else {
            assert!(!can_dereference(new_ptr));
        }
    }
}
