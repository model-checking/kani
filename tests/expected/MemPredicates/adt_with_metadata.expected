Failed Checks: Kani does not support reasoning about pointer to unallocated memory

Verification failed for - invalid_access::check_invalid_dyn_ptr
Complete - 3 successfully verified harnesses, 1 failures, 4 total.
